File: test/testBrokerBlocking.cpp
#include "broker.h"
#include "notice.h"

#include "./utility.h"

#include <gtest/gtest.h>
#include <pxr/pxr.h>
#include <pxr/usd/sdf/path.h>
#include <pxr/usd/usd/primRange.h>
#include <pxr/usd/usd/stage.h>

TEST(AddPrims, NoticesComparison)
{
    auto stage = PXR_NS::UsdStage::CreateInMemory();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction(
        [](const _Broker::StageNotice &){ return false; });

    ASSERT_TRUE(broker->IsInTransaction());

    // Edit the stage...
    stage->DefinePrim(PXR_NS::SdfPath {"/Foo"});
    stage->DefinePrim(PXR_NS::SdfPath {"/Bar"});
    stage->DefinePrim(PXR_NS::SdfPath {"/Baz"});

    // Ensure that broker notices are not sent during a transaction.
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);

    // While USD Notices are being sent as expected.
    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 6);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 3);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 3);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    ASSERT_FALSE(broker->IsInTransaction());

    // Ensure that no broker notices are sent after a transaction either.
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);
}

TEST(MuteLayers, NoticesComparison)
{
    auto stage = ::Test::CreateStageWithLayers();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction(
        [](const _Broker::StageNotice &){ return false; });

    ASSERT_TRUE(broker->IsInTransaction());

    // Edit the stage...
    auto layers = stage->GetRootLayer()->GetSubLayerPaths();

    // Keep ref pointer to the layer we try to mute and unmute to
    // prevent it for being destroyed when it is muted.
    auto _layer = PXR_NS::SdfLayer::FindOrOpen(layers[1]);

    stage->MuteLayer(layers[0]);
    stage->MuteLayer(layers[1]);
    stage->UnmuteLayer(layers[1]);
    stage->MuteAndUnmuteLayers(
        std::vector<std::string>{layers[2], layers[1]},
        std::vector<std::string>{});

    // Ensure that broker notices are not sent during a transaction.
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);

    // While USD Notices are being sent as expected.
    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 12);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 4);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 4);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 4);

    broker->EndTransaction();

    ASSERT_FALSE(broker->IsInTransaction());

    // Ensure that no broker notices are sent after a transaction either.
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);
}

TEST(ChangeEditTarget, NoticesComparison)
{
    auto stage = ::Test::CreateStageWithLayers();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    namespace _Broker = PXR_NS::UsdBrokerNotice;
    using _USD = PXR_NS::UsdNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        _Broker::StageNotice,
        _Broker::StageContentsChanged,
        _Broker::ObjectsChanged,
        _Broker::StageEditTargetChanged,
        _Broker::LayerMutingChanged
    > listener1(stage);

    // Subscribe to Usd Notices emitted by the stage.
    ::Test::Listener<
        _USD::StageNotice,
        _USD::StageContentsChanged,
        _USD::ObjectsChanged,
        _USD::StageEditTargetChanged,
        _USD::LayerMutingChanged
    > listener2(stage);

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction(
        [](const _Broker::StageNotice &){ return false; });

    ASSERT_TRUE(broker->IsInTransaction());

    // Edit the stage...
    auto layers = stage->GetRootLayer()->GetSubLayerPaths();
    auto layer1 = PXR_NS::SdfLayer::Find(layers[0]);
    auto layer2 = PXR_NS::SdfLayer::Find(layers[1]);

    stage->SetEditTarget(PXR_NS::UsdEditTarget(layer1));
    stage->SetEditTarget(PXR_NS::UsdEditTarget(layer2));

    // Ensure that broker notices are not sent during a transaction.
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);

    // While USD Notices are being sent as expected.
    ASSERT_EQ(listener2.Received<_USD::StageNotice>(), 2);
    ASSERT_EQ(listener2.Received<_USD::StageContentsChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::ObjectsChanged>(), 0);
    ASSERT_EQ(listener2.Received<_USD::StageEditTargetChanged>(), 2);
    ASSERT_EQ(listener2.Received<_USD::LayerMutingChanged>(), 0);

    broker->EndTransaction();

    ASSERT_FALSE(broker->IsInTransaction());

    // Ensure that no broker notices are sent after a transaction either.
    ASSERT_EQ(listener1.Received<_Broker::StageNotice>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageContentsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::ObjectsChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::StageEditTargetChanged>(), 0);
    ASSERT_EQ(listener1.Received<_Broker::LayerMutingChanged>(), 0);
}

TEST(CustomNotices, Handling)
{
    auto stage = PXR_NS::UsdStage::CreateInMemory();
    auto broker = PXR_NS::NoticeBroker::Create(stage);

    using namespace PXR_NS::UsdBrokerNotice;

    namespace _Broker = PXR_NS::UsdBrokerNotice;

    // Subscribe to notices emitted by the broker.
    ::Test::Listener<
        ::Test::MergeableNotice,
        ::Test::UnMergeableNotice
    > listener(stage);

    // Pass a predicate to block all broker notices.
    broker->BeginTransaction(
        [](const _Broker::StageNotice &){ return false; });

    ASSERT_TRUE(broker->IsInTransaction());

    broker->Send<::Test::MergeableNotice>(::Test::DataMap({{"Foo", "Test1"}}));
    broker->Send<::Test::MergeableNotice>(::Test::DataMap({{"Foo", "Test2"}}));
    broker->Send<::Test::MergeableNotice>(::Test::DataMap({{"Bar", "Test3"}}));

    broker->Send<::Test::UnMergeableNotice>();
    broker->Send<::Test::UnMergeableNotice>();
    broker->Send<::Test::UnMergeableNotice>();

    // Ensure that broker notices are not sent during a transaction.
    ASSERT_EQ(listener.Received<::Test::MergeableNotice>(), 0);
    ASSERT_EQ(listener.Received<::Test::UnMergeableNotice>(), 0);

    broker->EndTransaction();

    ASSERT_FALSE(broker->IsInTransaction());

    // Ensure that no broker notices are sent after a transaction either.
    ASSERT_EQ(listener.Received<::Test::MergeableNotice>(), 0);
    ASSERT_EQ(listener.Received<::Test::UnMergeableNotice>(), 0);
}

int main(int argc, char** argv)
{
    ::testing::InitGoogleTest(&argc, argv);
    return RUN_ALL_TESTS();
}

File: test/python/test_broker_blocking.py
# :coding: utf-8

import contextlib

from pxr import Usd, Sdf, Tf
from usd_notice_broker import NoticeBroker, BrokerNotice

import pytest


@contextlib.contextmanager
def listen_and_validate_notices(stage, notice_type, excepted_usd):
    """Edit stage and validate that USD notices are received
    while Broker notices are blocked.
    """
    broker = NoticeBroker.Create(stage)

    # Listen to broker notice.
    received_broker = []
    key1 = Tf.Notice.Register(
        getattr(BrokerNotice, notice_type),
        lambda n, _: received_broker.append(n), stage)

    # Listen to corresponding USD notice.
    received_usd = []
    key2 = Tf.Notice.Register(
        getattr(Usd.Notice, notice_type),
        lambda n, _: received_usd.append(n), stage)

    # Predicate blocking all broker notices.
    broker.BeginTransaction(predicate=lambda _: False)

    assert broker.IsInTransaction() is True

    # Edit the stage...
    yield

    # Ensure that broker notices are not sent during a transaction.
    assert len(received_broker) == 0

    # While USD Notices are being sent as expected.
    assert len(received_usd) == excepted_usd

    broker.EndTransaction()

    assert broker.IsInTransaction() is False

    # Ensure that no broker notices have been received.
    assert len(received_broker) == 0

@pytest.mark.parametrize("notice_type, excepted_usd", [
    ("StageNotice", 2),
    ("StageContentsChanged", 1),
    ("ObjectsChanged", 1),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_add_prim(notice_type, excepted_usd):
    """Add one prim to the stage and listen to notices.
    """
    stage = Usd.Stage.CreateInMemory()

    with listen_and_validate_notices(
        stage, notice_type, excepted_usd
    ):
        stage.DefinePrim("/Foo")

@pytest.mark.parametrize("notice_type, expected_usd", [
    ("StageNotice", 6),
    ("StageContentsChanged", 3),
    ("ObjectsChanged", 3),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_add_prims(notice_type, expected_usd):
    """Add several prims to the stage and listen to notices.
    """
    stage = Usd.Stage.CreateInMemory()

    with listen_and_validate_notices(
        stage, notice_type, expected_usd
    ):
        stage.DefinePrim("/Foo")
        stage.DefinePrim("/Bar")
        stage.DefinePrim("/Baz")

@pytest.mark.parametrize("notice_type, expected_usd", [
    ("StageNotice", 3),
    ("StageContentsChanged", 1),
    ("ObjectsChanged", 1),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 1),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_mute_layer(
    notice_type, expected_usd, stage_with_layers
):
    """Mute one layer and listen to notices.
    """
    stage = stage_with_layers

    with listen_and_validate_notices(
        stage, notice_type, expected_usd
    ):
        layers = stage.GetRootLayer().subLayerPaths
        stage.MuteLayer(layers[0])

@pytest.mark.parametrize("notice_type, expected_usd", [
    ("StageNotice", 12),
    ("StageContentsChanged", 4),
    ("ObjectsChanged", 4),
    ("StageEditTargetChanged", 0),
    ("LayerMutingChanged", 4),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_mute_layers(
    notice_type, expected_usd, stage_with_layers
):
    """Mute several layers and listen to notices.
    """
    stage = stage_with_layers

    with listen_and_validate_notices(
        stage, notice_type, expected_usd
    ):
        layers = stage.GetRootLayer().subLayerPaths

        # Keep ref pointer to the layer we try to mute and unmute to
        # prevent it for being destroyed when it is muted.
        _layer = Sdf.Layer.FindOrOpen(layers[1])

        stage.MuteLayer(layers[0])
        stage.MuteLayer(layers[1])
        stage.UnmuteLayer(layers[1])
        stage.MuteAndUnmuteLayers([layers[2], layers[1]], [])

@pytest.mark.parametrize("notice_type, expected_usd", [
    ("StageNotice", 2),
    ("StageContentsChanged", 0),
    ("ObjectsChanged", 0),
    ("StageEditTargetChanged", 2),
    ("LayerMutingChanged", 0),
], ids=[
    "StageNotice",
    "StageContentsChanged",
    "ObjectsChanged",
    "StageEditTargetChanged",
    "LayerMutingChanged",
])
def test_change_edit_target(
    notice_type, expected_usd, stage_with_layers
):
    """Mute several layers and listen to notices.
    """
    stage = stage_with_layers

    with listen_and_validate_notices(
        stage, notice_type, expected_usd
    ):
        layers = stage.GetRootLayer().subLayerPaths
        layer1 = Sdf.Layer.Find(layers[0])
        layer2 = Sdf.Layer.Find(layers[1])

        stage.SetEditTarget(Usd.EditTarget(layer1))
        stage.SetEditTarget(Usd.EditTarget(layer2))
